docs/cpp_build_systems.md

```diff
@@ -0,0 +1,350 @@
+
+## Design
+
+### Extracting information from Bazel
+
+Bazel's Starlark API provides [aspects](https://bazel.build/rules/aspects) to
+traverse the build graph, inspect build rules, define additional actions, and
+expose information through
+[providers](https://bazel.build/rules/rules#providers). For example, the
+`cc_proto_library` rule uses an aspect to traverse the dependency graph of
+`proto_library` rules, and dynamically attaches actions to generate C++ code
+using the Protobuf compiler and compile using the C++ compiler.
+
+In order to support multiple build systems, the goal is to define the overall
+build structure once for each system, and expose frequently-changing metadata
+from Bazel in a way that can be included from the build definition. Primarily,
+this means exposing the list of source files in a way that they can be included
+in other build definitions.
+
+Two aspects are used to extract this information from the Bazel build
+definitions:
+
+*   `cc_file_list_aspect` extracts `srcs`, `hdrs`, and `textual_hdrs` from build
+    rules like `cc_library`. The sources are exposed through a provider named
+    `CcFileList`.
+*   `proto_file_list_aspect` extracts the `srcs` from a `proto_library`, and
+    also generates the expected filenames that would be generated by the
+    Protobuf compiler. This information is exposed through a provider named
+    `ProtoFileList`.
+
+On their own, these aspects have limited utility. However, they can be
+instantiated by custom rules, so that an ordinary `BUILD.bazel` target can
+produce outputs based on the information gleaned from these aspects.
+
+### (Aside) Distribution libraries
+
+Bazel's native `cc_library` rule is typically used on a "fine-grained" level, so
+that, for example, lightweight unit tests can be written with narrow scope.
+Although Bazel does build library artifacts (such as `.so` and `.a` files on
+Linux), they correspond to `cc_library` rules.
+
+Since the entire "Protobuf library" includes several constituent `cc_library`
+files, a special rule, `cc_dist_library`, combines several fine-grained
+libraries into a single, monolithic library.
+
+For the Protobuf project, these "distribution libraries" are intended to match
+the granularity of the Autotools- and CMake-based builds. Since the Bazel-built
+distribution library covers the rules with the source files needed by other
+builds, the `cc_dist_library` rule invokes the `cc_file_list_aspect` on its
+input libraries. The result is that a `cc_dist_library` rule not only produces
+composite library artifacts, but also collect and provide the list of sources
+that were inputs.
+
+For example:
+
+```
+$ cat cc_dist_library_example/BUILD.bazel 
+load("@rules_cc//cc:defs.bzl", "cc_library")
+load("//pkg:cc_dist_library.bzl", "cc_dist_library")
+
+cc_library(
+    name = "a",
+    srcs = ["a.cc"],
+)
+
+cc_library(
+    name = "b",
+    srcs = ["b.cc"],
+)
+
+cc_dist_library(
+    name = "c",
+    deps = [
+        ":a",
```

Comment:
These are good questions. There are some comments in the rule implementation that address some of this, but they probably don't address this well enough...

> Does anything alert us if a cc_library target is missing from this list?

In the future state where e.g. cmake file lists get regenerated on the fly, then the cmake tests would give some coverage. Other than that, there isn't a great mechanism. (More comments below, though...)

> Is there every any case where we'd want to exclude some?

Yes, there are a couple of cases I can think of:

1. Make the "full" library depend on the "lite" library. Currently, for the non-Bazel builds, the full `libprotobuf` library contains all of `libprotobuf_lite`. This probably isn't the right thing if there are mixed transitive dependencies on both.
2. Exclude transitive dependencies, like zlib. A system-installable library should use other system-installed libraries.


> If not, could we wrap cc_library to also add itself here automatically?

Here is a discussion of that in the rule implementation, if you are interested:
https://github.com/protocolbuffers/protobuf/blob/main/pkg/cc_dist_library.bzl#L316-L334

Basically, the `cc_import` rule could be used, but there are problems with making a drop-in that Bazel would be happy with, e.g., what headers are exposed and how. I think there is also some value in explicitly separating the archive/static library from the DSO, so defining two separate `cc_import`s would be a much more natural way to do that. I think this is doable, it just requires more design thought.

If the composite dist libraries can look more like normal `cc_library`s, then there is probably some value in making them usable instead of the main runtime targets.

(Continuing in part from your first question...) For example, the tests could depend on the main runtime conditionally, either through the ordinary `cc_library` targets or the `cc_dist_library` targets. Linking the tests against the dist targets would be another way to verify correctness of their definitions.

On top of all of this, there is already some experimental work in Bazel for building composite libraries like this, so there is some risk of a separate implementation not quite doing the right thing.
https://github.com/bazelbuild/bazel/blob/master/src/main/starlark/builtins_bzl/common/cc/experimental_cc_shared_library.bzl

Anyhow, I think the short answer is: no, we can't use `cc_library` or `cc_import` as-is. There probably is some value in exposing the composite libraries through the Bazel build, but gaining the benefits will require some careful design thought.